interop-testing/src/main/java/io/grpc/testing/integration/XdsTestClient.java

```diff
@@ -197,65 +268,87 @@ public void run() {
+                .withInterceptors(
+                    new ClientInterceptor() {
+                      @Override
+                      public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(
+                          MethodDescriptor<ReqT, RespT> method,
+                          CallOptions callOptions,
+                          Channel next) {
+                        ClientCall<ReqT, RespT> call = next.newCall(method, callOptions);
+                        clientCallRef.set(call);
+                        return new SimpleForwardingClientCall<ReqT, RespT>(call) {
+                          @Override
+                          public void start(Listener<RespT> responseListener, Metadata headers) {
+                            headers.merge(headersToSend);
+                            super.start(
+                                new SimpleForwardingClientCallListener<RespT>(responseListener) {
+                                  @Override
+                                  public void onHeaders(Metadata headers) {
+                                    hostnameRef.set(headers.get(XdsTestServer.HOSTNAME_KEY));
+                                    super.onHeaders(headers);
+                                  }
+                                },
+                                headers);
+                          }
+                        };
+                      }
+                    });
 
-              @Override
-              public void onMessage(SimpleResponse response) {
-                hostname = response.getHostname();
-                // TODO(ericgribkoff) Currently some test environments cannot access the stats RPC
-                // service and rely on parsing stdout.
-                if (printResponse) {
-                  System.out.println(
-                      "Greeting: Hello world, this is "
-                          + hostname
-                          + ", from "
-                          + call.getAttributes().get(Grpc.TRANSPORT_ATTR_REMOTE_ADDR));
-                }
-              }
-
-              @Override
-              public void onClose(Status status, Metadata trailers) {
-                if (printResponse && !status.isOk()) {
-                  logger.log(Level.WARNING, "Greeting RPC failed with status {0}", status);
-                }
-                for (XdsStatsWatcher watcher : savedWatchers) {
-                  watcher.rpcCompleted(requestId, hostname);
-                }
-              }
-            },
-            new Metadata());
-
-        call.sendMessage(request);
-        call.request(1);
-        call.halfClose();
+        if (rpcType == RpcType.EMPTY_CALL) {
+          stub.emptyCall(EmptyProtos.Empty.getDefaultInstance());
+        } else if (rpcType == RpcType.UNARY_CALL) {
+          SimpleRequest request = SimpleRequest.newBuilder().setFillServerId(true).build();
+          SimpleResponse response = stub.unaryCall(request);
```

Comment:
Is there an easy way for the client to read both resp.Hostname and the hostname from header?

This will be necessary if we want to merge this PR before updating the test server image to send hostname in headers.

Reply:
Easy enough but it will make the code a bit cluttered and doesn't seem necessary - the server will be backwards compatible itself so we can just update the server image prior to merging, right? If that's more complicated than I think, then I can add a fallback here with a TODO to remove it.

Reply:
Nevermind, added the fallback